### PR TITLE
[configurator] Return pkg-config error to clients of new query API.

### PR DESCRIPTION
**Pull request by ejgallego**
It is very convenient for configuration tools to be able to provide a
better error message to the user when pkg-config fails.

This was suggested by @garrigue.

We take advantage of the newly [undocumented] API in 1.7.2 to modify
its return type, but this could be a problem. If so, we will have to
introduce yet another function if we want this functionality.

### CODE REVIEW

File: src/configurator/v1.mli
Comment (ejgallego) [inline src/configurator/v1.mli]: Umm, exposing `Stdune` this in this interface file may be a problem right?

Comment (rgrinberg) [reply src/configurator/v1.mli]: Yes, that's definitely not acceptable. Why don't you just return the result type? This will mean that this will require the result compatibility package in configurator, but I think that's acceptable.

**Comment by rgrinberg** [comment]
Are we allowed to make a breaking change like this to `query_expr`? I think we probably are as you're the only consumer, but we need to double check somehow.

**Comment by ejgallego** [comment]
> Are we allowed to make a breaking change like this to `query_expr`? I think we probably are as you're the only consumer, but we need to double check somehow.

I guess a way to check is to see if there are any `dune >= 1.7.2` dependencies in OPAM.

File: src/configurator/v1.mli
Comment (ejgallego) [reply src/configurator/v1.mli]: Umm, how should such package be exported without exposing `Stdune`?

**Comment by diml** [comment]
BTW, in general we should a bit be more strict about minor releases. I.e. we shouldn't introduce new functionalities in minor releases, just to give new functionalities a bit more time to mature in master.

**Comment by rgrinberg** [comment]
> BTW, in general we should a bit be more strict about minor releases. I.e. we shouldn't introduce new functionalities in minor releases, just to give new functionalities a bit more time to mature in master.

Yeah, that was definitely a mistake. I originally framed this issue as a "regression" so the new API was supposed to be a fix. I'll be more careful.

File: src/configurator/v1.mli
Comment (rgrinberg) [reply src/configurator/v1.mli]: You did exactly what I meant. The result type in stdune matches the one in stdlib so there's no need to export it from stdune.

**Comment by rgrinberg** [comment]
@ejgallego you're getting a compilation error because of a bug on our end. Because we don't have transitive dependencies anymore, `Dune_caml`'s cmi's don't get included.

Just apply this patch:
```
diff --git a/src/configurator/dune b/src/configurator/dune
index 0507edec..697f264d 100644
--- a/src/configurator/dune
+++ b/src/configurator/dune
@@ -3,5 +3,5 @@
 (library
  (name        configurator)
  (public_name dune.configurator)
- (libraries   stdune ocaml_config dune_lang)
+ (libraries   stdune ocaml_config dune_lang dune_caml)
  (flags       (:standard -safe-string (:include flags/flags.sexp))))
```

And things should be fixed. In fact, we should depend on `dune_caml` anywhere we depend on `stdune`. @diml do we need to think of a solution for re-exporting libraries? I believe @bobot brought this issue up before.

**Comment by ejgallego** [comment]
> @ejgallego you're getting a compilation error because of a bug on our end. Because we don't have transitive dependencies anymore, `Dune_caml`'s cmi's don't get included.

Ok, I see thanks! I was indeed puzzled by the error, but now I understand it.

**Comment by ejgallego** [comment]
> Ok, I see thanks! I was indeed puzzled by the error, but now I understand it.

This doesn't work in 4.02 [as expected] so which compat module should we use for it? Should we open `Result`?

**Comment by rgrinberg** [comment]
I think you can just use `Dune_caml.result` directly in the mli. Pretty sure we can make `Dune_caml` public as it doesn't actually add new api's. But let's see what @diml says.

**Comment by ejgallego** [comment]
> I think you can just use `Dune_caml.result` directly in the mli. Pretty sure we can make `Dune_caml` public as it doesn't actually add new api's. But let's see what @diml says.

Let's go for it to try to get a full CI build.

**Comment by diml** [comment]
Hmm, I'm not a big fan of publicising `Dune_caml` given that it is not complete (it doesn't export all the stdlib modules) and we will get rid of it once we drop support for OCaml < 4.07. I suggest to put the result stuff in its own `dune.result` library, or just define a new result type in configurator. The latter is probably the simplest solution.

**Comment by ejgallego** [comment]
> The latter is probably the simplest solution.

Indeed that seems to make the most sense, I've added a result type that can be deprecated and made into alias once 4.02 support is gone.

I did grep for all the dependencies on dune in opam-repository and indeed there is no package depending on `dune >= 1.7.2` that I could find, so tweaking this bit of API IMHO can be done, tho personally I would just do another 1.7 release removing the newly-added API and defer this new function to 1.8

**Comment by rgrinberg** [comment]
> tho personally I would just do another 1.7 release removing the newly-added API and defer this new function to 1.8

This seems like a case where the cure is worse than the disease. Users might easily introduce issues to their code by having it work against 1.7.2.

There's nothing fundamentally wrong with the old function, perhaps we can just think of a new name for the function that returns a result? `query_expr_err` sounds find to me for example. Then we can deprecate the old function if it's really necessary. Although I don't see such a strong need as it doesn't introduce any problems.

**Comment by rgrinberg** [comment]
> Hmm, I'm not a big fan of publicising `Dune_caml` given that it is not complete (it doesn't export all the stdlib modules) and we will get rid of it once we drop support for OCaml < 4.07. I suggest to put the result stuff in its own `dune.result` library, or just define a new result type in configurator. The latter is probably the simplest solution.

Well we'll get rid of `dune.result` even earlier as it will not be useful once we drop support for 4.02.3.

In any case, I like the dune.result idea. Will implement it in a separate PR.

**Comment by rgrinberg** [comment]
This pr now needs https://github.com/ocaml/dune/pull/1892

**Comment by ejgallego** [comment]
> There's nothing fundamentally wrong with the old function, perhaps we can just think of a new name for the function that returns a result? `query_expr_err` sounds find to me for example. Then we can deprecate the old function if it's really necessary. Although I don't see such a strong need as it doesn't introduce any problems.

Sure, I'll rename and add another function.

IMO I would deprecate the old one right away, no reason to keep the API creep.

> This pr now needs #1892

Cool, I'll wait for it to be solved.

**Comment by ejgallego** [comment]
Umm, I am not sure I understand how to use the new lib.

**Comment by rgrinberg** [comment]
I believe that this is all you need:

```
diff --git a/src/configurator/dune b/src/configurator/dune
index c122bdbc..14efb6fc 100644
--- a/src/configurator/dune
+++ b/src/configurator/dune
@@ -3,5 +3,5 @@
 (library
  (name        configurator)
  (public_name dune.configurator)
- (libraries   stdune ocaml_config dune_lang dune._result)
+ (libraries   stdune ocaml_config dune_lang dune_caml dune._result)
  (flags       (:standard -safe-string (:include flags/flags.sexp))))
diff --git a/src/configurator/v1.mli b/src/configurator/v1.mli
index b56ffcff..cadae2fc 100644
--- a/src/configurator/v1.mli
+++ b/src/configurator/v1.mli
@@ -1,3 +1,5 @@
+open! Dune_result
+
 type t
 
 val create
```

**Comment by ejgallego** [comment]
> (libraries   stdune ocaml_config dune_lang dune_caml dune._result)

Oh indeed thanks, I was missing the `dune_caml` lib as for the compiler to be able to spot the transitive equality on types.

File: src/configurator/v1.ml
Comment (rgrinberg) [inline src/configurator/v1.ml]: This is a standard function, so let's just add this to `Stdune.Result`. In base, this is called `Or_error.ok`, but that's probably too cute. Let's just call it `Result.to_option` or something.

**Review by rgrinberg — APPROVED** [review]
Looks good. Just one minor nit. We prefer to keep all the general purpose functions in stdune.

File: src/configurator/v1.ml
Comment (ejgallego) [reply src/configurator/v1.ml]: Indeed; I didn't add this to stdune I guess to keep it 1:1 to OCaml's result, but that's the wrong reason. Do you want me to update it on this PR or would you like it to introduce the change in a separate one?

Comment (rgrinberg) [reply src/configurator/v1.ml]: In stdune we extend the stdlib freely btw. We don't even bother to maintain compatibility. Dune_caml is the 1:1 "namespaced" clone, which we only need because `Stdlib` is >= 4.07. Feel free to just do it here, in a new PR, or just push to master. This is a very a minor change.

Comment (ejgallego) [reply src/configurator/v1.ml]: Ok will do it here as this PR is the first consumer of such a function.

**Comment by ejgallego** [comment]
Review re-requested; main point to discuss could be the deprecation of the 1.7.2 introduced function, otherwise this should be OK I think.

**Comment by rgrinberg** [comment]
The deprecation is acceptable to me.

**Review by emillon — APPROVED** [review]
I don't think we have to deprecate the function right away, but I'm fine if we do.

**Comment by ejgallego** [comment]
> I don't think we have to deprecate the function right away, but I'm fine if we do.

It is not used by anyone yet, so IMO it is better if we just try to avoid people use it at all as to have a more compact API.

By the way, maybe at some point it would be useful for dune to include some kind of "dune-config" script? For example https://github.com/garrigue/lablgtk/blob/lablgtk3/src/dune_config.ml

This allows people to write:
```
(rule
 (targets
   cflag-gtk+-3.0.sexp
   clink-gtk+-3.0.sexp)
 (action (run ./dune_config.exe -pkg gtk+-3.0 -version 3.18)))
...
 (c_flags         (:include cflag-gtk+-3.0.sexp))
 (c_library_flags (:include clink-gtk+-3.0.sexp))
```
it seems to me that this is enough of a standard pattern as to avoid having N config-like scripts.

**Comment by diml** [comment]
At this point we might as well add a `(pkg_config gtk+-3.0)` field

**Comment by ejgallego** [comment]
> At this point we might as well add a `(pkg_config gtk+-3.0)` field

IMO that'd be a very nice solution and cover a lot of cases, let me open a ticket for it.

**Comment by rgrinberg** [comment]
@ejgallego btw, we usually merge our own PR's after approval. So if you're waiting for someone else to merge it, it's unlikely to happen.

**Comment by ejgallego** [comment]
Indeed, sorry. I'm used to the coq model where you never merge your own PRs.

**Comment by rgrinberg** [comment]
I merged this myself to unblock 1.8

**Pull request closed without merging**
